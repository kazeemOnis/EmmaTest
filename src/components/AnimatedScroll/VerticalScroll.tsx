import React from 'react';
import {Dimensions, FlatList} from 'react-native';
import {Content} from '../../components';
import {Data} from '../../types';

interface AnimatedScrollProps {
  data: Data[];
  activeIndex: number;
  activeScroll: 'vertical' | 'horizontal';
  setActiveIndex: React.Dispatch<React.SetStateAction<number>>;
  setActiveScroll: React.Dispatch<
    React.SetStateAction<'vertical' | 'horizontal'>
  >;
}

export const VerticalScroll = ({
  data,
  activeIndex,
  activeScroll,
  setActiveIndex,
  setActiveScroll,
}: AnimatedScrollProps) => {
  const ref = React.useRef<FlatList>(null);

  React.useEffect(() => {
    if (activeScroll === 'horizontal') {
      ref.current?.scrollToIndex({
        index: activeIndex,
        animated: true,
      });
    }
  }, [activeScroll, activeIndex]);

  return (
    <FlatList
      ref={ref}
      data={data}
      initialScrollIndex={activeIndex}
      showsVerticalScrollIndicator={false}
      snapToAlignment="start"
      decelerationRate="fast"
      snapToInterval={Dimensions.get('window').height}
      onResponderMove={_event => {
        setActiveScroll('vertical');
      }}
      onScroll={event => {
        const index = Math.abs(
          Math.round(
            event.nativeEvent.contentOffset.y / Dimensions.get('window').height,
          ),
        );
        if (activeScroll === 'vertical') {
          setActiveIndex(index);
        }
      }}
      renderItem={({item}: {item: Data}) => (
        <Content
          title={item.title}
          subtext={item.subtext}
          content={item.content}
        />
      )}
    />
  );
};
