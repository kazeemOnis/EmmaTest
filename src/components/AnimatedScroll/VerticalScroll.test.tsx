import React from 'react';
import {cleanup, render} from '@testing-library/react-native';
import {act} from 'react-test-renderer';
import {VerticalScroll} from './VerticalScroll';

afterEach(() => {
  cleanup();
});

const mockSetActiveIndex = jest.fn();
const mockSetActiveScroll = jest.fn();

const mockData = {
  data: [
    {
      id: 1,
      title: 'Kazeem',
      subtext: 'React Native Engineer',
      image: 'https://images.unsplash.com/photo-1579353977828-2a4eab540b9a',
      content:
        'Experienced React Native Engineer with over 4 years of experience in building and shipping production ready apps to users',
    },
  ],
};

const setup = (overrides = {}) => {
  const props = {...mockData, ...overrides};
  const utils = render(
    <VerticalScroll
      {...props}
      activeIndex={0}
      activeScroll="horizontal"
      setActiveIndex={mockSetActiveIndex}
      setActiveScroll={mockSetActiveScroll}
    />,
  );
  return {...utils, props};
};

describe('Vertical Scroll', () => {
  it('Should render vertical scroll', async () => {
    const {toJSON} = setup();
    await act(async () => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
