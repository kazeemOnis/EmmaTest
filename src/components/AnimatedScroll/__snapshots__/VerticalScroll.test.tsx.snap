// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vertical Scroll Should render vertical scroll 1`] = `
<RCTScrollView
  data={
    [
      {
        "content": "Experienced React Native Engineer with over 4 years of experience in building and shipping production ready apps to users",
        "id": 1,
        "image": "https://images.unsplash.com/photo-1579353977828-2a4eab540b9a",
        "subtext": "React Native Engineer",
        "title": "Kazeem",
      },
    ]
  }
  decelerationRate="fast"
  getItem={[Function]}
  getItemCount={[Function]}
  initialScrollIndex={0}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onResponderMove={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  showsVerticalScrollIndicator={false}
  snapToAlignment="start"
  snapToInterval={1334}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          [
            {
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "height": 1334,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "white",
                "fontFamily": "Nunito",
                "fontSize": 24,
                "marginBottom": 4,
                "textAlign": "center",
              },
            ]
          }
          testID="contentTitle"
        >
          Kazeem
        </Text>
        <Text
          style={
            [
              {
                "color": "white",
                "fontFamily": "Nunito",
                "fontSize": 16,
                "textAlign": "center",
              },
            ]
          }
          testID="contentSubtext"
        >
          React Native Engineer
        </Text>
        <Text
          style={
            [
              {
                "color": "white",
                "fontFamily": "Nunito",
                "fontSize": 14,
                "marginTop": 16,
              },
            ]
          }
          testID="contentContent"
        >
          Experienced React Native Engineer with over 4 years of experience in building and shipping production ready apps to users
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
